### 🏎️ Task 2.3: Matching Hassan to the Right F1 Team

We aim to find the best-fitting F1 team for **Hassan**, a promising driver, by comparing his driving traits to team philosophies.

#### 🔍 Objective:

* Quantify Hassan's compatibility with **Red Bull**, **Ferrari**, and **Mercedes**.
* Use vector-based similarity and distance metrics.

#### 📊 Methods Used:

* **Cosine Similarity** – measures directional alignment.
* **Euclidean Distance** – measures straight-line similarity.
* **Manhattan Distance** – sums trait-by-trait differences.
* **Pearson Correlation** – Measures linear correlation.
* **Spearman Correlation** – Measures monotonic relationship. 

#### 🏁 Goal:

* Recommend the team with the **highest statistical and cultural alignment** to Hassan’s profile.

### `imports`

In [ ]:
import numpy as np
from scipy.spatial.distance import cosine, euclidean, cityblock  # all are distances
from scipy.stats import pearsonr, spearmanr  # correlation coefficients

In [4]:
# Hassan's profile
hassan = np.array([9, 8, 7, 6, 7, 8, 6])

In [ ]:
# Team profiles
red_bull = np.array([10, 9, 6, 7, 6, 9, 5])
ferrari  = np.array([9, 7, 6, 6, 7, 7, 5])
mercedes = np.array([8, 6, 8, 9, 9, 5, 9])

**`Cosine Similarity`**

In [6]:
# define cosine similarity function
def cosine_similarity(a, b): # scipy function returns cosine distance ( cosine distance = 1−cos(θ) ), lower is better
    return 1 - cosine(a, b)  # as required is cosine similarity (higher is better)

In [7]:
cosine_results = {
    "Red Bull": cosine_similarity(hassan, red_bull),
    "Ferrari": cosine_similarity(hassan, ferrari),
    "Mercedes": cosine_similarity(hassan, mercedes)
}

In [27]:
cosine_results

{'Red Bull': 0.9917790279239155,
 'Ferrari': 0.9972562318087912,
 'Mercedes': 0.9564218442002067}

* **Explanation**: Measures the **angle** between two vectors. Values range **from -1** (opposite, theta=180) **to +1** (identical, theta=0).
* **Result**: Ferrari **(0.9974)** has the highest similarity to Hassan, indicating strong alignment in direction of performance traits.
* ✅ **Best Fit**: **Ferrari**

**`Euclidean Distance`**

In [ ]:
def euclidean_distance(a, b):  # any distance, lower is better
    return euclidean(a, b)

In [24]:
euclidean_results = {
    "Red Bull": euclidean_distance(hassan, red_bull),
    "Ferrari": euclidean_distance(hassan, ferrari),
    "Mercedes": euclidean_distance(hassan, mercedes)
}

In [26]:
euclidean_results

{'Red Bull': 2.6457513110645907, 'Ferrari': 2.0, 'Mercedes': 6.082762530298219}

* **Explanation**: Measures the **straight-line** distance between two vectors. Smaller distance → more similar.
* **Result**: Ferrari **(2.0)** is closest to Hassan.
* ✅ **Best Fit**: **Ferrari**

**`Manhattan Distance`**

In [ ]:
def manhattan_distance(a, b):   # absolute distance, lower is better
    return cityblock(a, b)

In [29]:
manhattan_results = {
    "Red Bull": manhattan_distance(hassan, red_bull),
    "Ferrari": manhattan_distance(hassan, ferrari),
    "Mercedes": manhattan_distance(hassan, mercedes)
}

In [30]:
manhattan_results

{'Red Bull': 7, 'Ferrari': 4, 'Mercedes': 15}

* **Explanation**: Measures the **sum of absolute differences** across all traits. Simpler but effective.
* **Result**: Ferrari **(4)** again has the lowest distance.
* ✅ **Best Fit**: **Ferrari**

**`Pearson Correlation`**

In [42]:
def pearson_correlation(a, b):  # measures liner correlation, higher is better
    return pearsonr(a, b)[0]   # returns ony correlation coefficient

In [43]:
pearson_results = {
    "Red Bull": pearson_correlation(hassan, red_bull),
    "Ferrari": pearson_correlation(hassan, ferrari),
    "Mercedes": pearson_correlation(hassan, mercedes)
}

In [44]:
pearson_results

{'Red Bull': 0.8773464777984028,
 'Ferrari': 0.9046950831485676,
 'Mercedes': -0.6004805767690768}

* **Explanation**: Measures **linear correlation**. Values close to 1 mean similar trend/pattern.
* **Result**: Ferrari **(0.9047)** again shows the highest alignment.
* ✅ **Best Fit**: **Ferrari**

**`Spearman Correlation`**

In [36]:
def spearman_correlation(a, b):  # measures monotonic relationship, higher is better
    return spearmanr(a, b)[0]    # returns coefficient

In [37]:
spearman_results = {
    "Red Bull": spearman_correlation(hassan, red_bull),
    "Ferrari": spearman_correlation(hassan, ferrari),
    "Mercedes": spearman_correlation(hassan, mercedes)
}

In [38]:
spearman_results

{'Red Bull': 0.8411582311380665,
 'Ferrari': 0.9040133315325639,
 'Mercedes': -0.7597558850114101}

* **Explanation**: Measures **rank correlation**, assessing how well the relationship between two sets of traits can be described by a **monotonic trend** (not necessarily linear). It compares the **order** of values, not their magnitude.
* **Result**: Ferrari **(0.9040)** showw the highest rank alignment with Hassan’s traits.
* ✅ **Best Fit**: **Ferrari**

## 🏁 Final Recommendation

Based on **all similarity metrics (Cosine, Euclidean, Manhattan, Pearson, Spearman)**, **Ferrari Racing** is the most compatible team for Hassan. Their aggressive and pace-oriented philosophy closely mirrors Hassan’s explosive driving style, adaptability, and high team engagement.

> ✅ **Recommended Team: Ferrari Racing**